maven-embedder/src/test/java/org/apache/maven/cli/MavenCliTest.java

```diff
@@ -552,6 +553,38 @@ public void findRootProjectWithAttribute() {
         assertEquals(test, new DefaultRootLocator().findRoot(test.resolve("child")));
     }
 
+    @Test
+    public void testPropertiesInterpolation() throws Exception {
+        // Arrange
+        CliRequest request = new CliRequest(
+                new String[] {
+                    "-Dfoo=bar",
+                    "-DvalFound=s${foo}i",
+                    "-DvalNotFound=s${foz}i",
+                    "-DvalRootDirectory=${session.rootDirectory}/.mvn/foo",
+                    "-DvalTopDirectory=${session.topDirectory}/pom.xml",
+                    "-f",
+                    "${session.rootDirectory}/my-child",
+                    "prefix:3.0.0:${foo}",
+                    "validate"
```

Comment:
As far as I understand, they need to be defined ahead of time, right?

Reply:
I mean `foo`. You cannot foo before it has been defined, just like symbols in C. It does not perform a late eval. So is the following possible?

```
-Dbla=${foo}
-Dfoo=blubb
```

Reply:
They (user properties) are first gathered, then interpolated. So yes, it is possible. Next, using interpolated user properties the rest of CLI is interpolated.